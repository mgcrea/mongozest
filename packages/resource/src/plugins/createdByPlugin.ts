import { DefaultSchema } from '@mongozest/core';
import { ObjectId } from 'mongodb';
import { Resource } from '../resource';

export type CreatedByPluginSchema = {
  createdBy: ObjectId;
  updatedBy: ObjectId;
};

export const createdByPlugin = <
  TSchema extends DefaultSchema & CreatedByPluginSchema = DefaultSchema & CreatedByPluginSchema
>(
  resource: Resource<TSchema>,
  { idKey = '_id', createdByKey = 'createdBy', updatedByKey = 'updatedBy' } = {}
): void => {
  resource.pre('postCollection', (operation) => {
    const document = operation.get('document');
    const req = operation.get('request');
    if (req.user && req.user[idKey as '_id']) {
      document[createdByKey as 'createdBy'] = req.user[idKey as '_id'];
      document[updatedByKey as 'updatedBy'] = req.user[idKey as '_id'];
    }
  });
  resource.pre('patchCollection', (operation, _filter, update) => {
    const req = operation.get('request');
    if (req.user && req.user[idKey as '_id']) {
      if (!update.$set) {
        update.$set = {};
      }
      Object.assign(update.$set, { [updatedByKey as 'updatedBy']: req.user[idKey as '_id'] });
    }
  });
  resource.pre('patchDocument', (operation, _filter, update) => {
    const req = operation.get('request');
    if (req.user && req.user[idKey as '_id']) {
      if (!update.$set) {
        update.$set = {};
      }
      Object.assign(update.$set, { [updatedByKey as 'updatedBy']: req.user[idKey as '_id'] });
    }
  });
};
