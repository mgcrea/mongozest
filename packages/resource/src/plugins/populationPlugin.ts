import { DefaultSchema } from '@mongozest/core';
import { Request } from 'express';
import createError from 'http-errors';
import JSON5 from 'json5';
import { isString, mapValues, pick } from 'lodash';
import 'mongodb';
import { FilterQuery, FindOneOptions, SchemaMember } from 'mongodb';
import { OperationMap } from '../operation';
import { Resource } from '../resource';

export type Population<T> = SchemaMember<T, number | boolean | any>;

declare module 'mongodb' {
  interface FindOneOptions<T> {
    population?: Population<T>;
  }
}

export const populatePlugin = <TSchema extends DefaultSchema = DefaultSchema>(
  resource: Resource<TSchema>,
  { strictJSON = false, optionName = 'population' } = {}
): void => {
  const parseQueryParam = (value: any, key: string) => {
    if (!isString(value) || !/^[\[\{]/.test(value)) {
      return value;
    }
    try {
      return (strictJSON ? JSON : JSON5).parse(value);
    } catch (err) {
      throw createError(400, `Failed to parse query field=\`${key}\``);
    }
  };
  const preparePopulation = (
    operation: OperationMap<TSchema>,
    _filter: FilterQuery<TSchema>,
    options: FindOneOptions<TSchema>
  ) => {
    const req: Request = operation.get('request');
    const whitelist = [optionName];
    const queryOptions = mapValues(pick(req.query, whitelist), parseQueryParam);
    if (queryOptions[optionName]) {
      options[optionName as 'population'] = queryOptions[optionName];
    }
  };
  resource.pre('getCollection', preparePopulation);
  resource.pre('getDocument', preparePopulation);
  resource.pre('patchDocument', (operation, filter, _update, options) => preparePopulation(operation, filter, options));
};
