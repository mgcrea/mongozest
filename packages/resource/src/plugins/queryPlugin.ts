import {DefaultSchema} from '@mongozest/core';
import createError from 'http-errors';
import JSON5 from 'json5';
import {isEmpty, isString, mapValues, pick} from 'lodash';
import {Resource} from '../resource';

export const queryPlugin = <TSchema extends DefaultSchema = DefaultSchema>(
  resource: Resource<TSchema>,
  {strictJSON = false} = {}
): void => {
  const parseQueryParam = (value: any, key: string) => {
    if (!isString(value) || !/^[\[\{]/.test(value)) {
      return value;
    }
    try {
      return (strictJSON ? JSON : JSON5).parse(value);
    } catch (err) {
      throw createError(400, `Failed to parse query field=\`${key}\``);
    }
  };
  const castQueryParam = (value: any, key: string) => {
    switch (key) {
      case 'limit':
      case 'skip':
      case 'min':
      case 'max':
        return value * 1;
      default:
        return value;
    }
  };

  resource.pre('filter', (operation) => {
    const req = operation.get('request');
    const filter = operation.get('filter');
    const queryFilter = parseQueryParam(req.query.filter, 'filter');
    if (queryFilter) {
      const nextFilter = !isEmpty(filter) ? {$and: [filter, queryFilter]} : queryFilter;
      operation.set('filter', nextFilter);
    }
  });

  resource.pre('getCollection', (operation, _filter, options) => {
    const req = operation.get('request');
    const whitelist = ['projection', 'limit', 'min', 'max', 'sort', 'skip'];
    const queryOptions = mapValues(mapValues(pick(req.query, whitelist), parseQueryParam), castQueryParam);
    Object.assign(options, queryOptions);
  });

  resource.pre('patchCollection', (operation, _filter, _update, options) => {
    const req = operation.get('request');
    const whitelist = ['projection'];
    const queryOptions = mapValues(mapValues(pick(req.query, whitelist), parseQueryParam), castQueryParam);
    Object.assign(options, queryOptions);
  });

  resource.pre('getDocument', (operation, _filter, options) => {
    const req = operation.get('request');
    const whitelist = ['projection'];
    const queryOptions = mapValues(mapValues(pick(req.query, whitelist), parseQueryParam), castQueryParam);
    Object.assign(options, queryOptions);
  });

  resource.pre('patchDocument', (operation, _filter, _update, options) => {
    const req = operation.get('request');
    const whitelist = ['projection'];
    const queryOptions = mapValues(mapValues(pick(req.query, whitelist), parseQueryParam), castQueryParam);
    Object.assign(options, queryOptions);
  });
};
