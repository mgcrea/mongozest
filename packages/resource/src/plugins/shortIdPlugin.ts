import {Resource} from '..';

const SID_REGEX = /^[a-z0-9\-\_]{7,14}$/i;

export default function shortIdPlugin(resource: Resource, {sidKey = '_sid'} = {}) {
  // const handle = (key: string, value: string) => {

  // };

  // resource.addUrl((key, value) => {

  // })

  resource.addIdentifierHandler(
    (_sid: string) => SID_REGEX.test(_sid),
    (_sid: string) => ({_sid})
  );

  // resource.pre('id', (key: string, value: string, filter: {[s: string]: any}) => {
  //   const isMatch = key === '0' && SID_REGEX.test(value);
  //   return isMatch ? {[sidKey]: toString(value)} : null;
  // });
}
