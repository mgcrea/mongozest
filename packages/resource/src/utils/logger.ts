export { inspect, log } from '@mongozest/core';
