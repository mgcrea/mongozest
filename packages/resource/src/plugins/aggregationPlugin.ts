import {AggregationPipeline, DefaultSchema} from '@mongozest/core';
import {Request, Response, Router} from 'express';
import createError from 'http-errors';
import JSON5 from 'json5';
import {isEmpty, isString} from 'lodash';
import {CollectionAggregationOptions, FilterQuery} from 'mongodb';
import {createOperationMap} from '../operation';
import {Resource} from '../resource';
import {asyncHandler} from '../utils/request';

export const aggregationPlugin = <TSchema extends DefaultSchema = DefaultSchema>(
  resource: Resource<TSchema>,
  {strictJSON = false, pipelineParamName = 'pipeline', pathName = 'aggregate'} = {}
): void => {
  const parseQueryParam = (value: any, key: string) => {
    if (!isString(value) || !/^[\[\{]/.test(value)) {
      return value;
    }
    try {
      return (strictJSON ? JSON : JSON5).parse(value);
    } catch (err) {
      throw createError(400, `Failed to parse query field=\`${key}\``);
    }
  };
  const castQueryParam = (value: any, key: string) => {
    switch (key) {
      case pipelineParamName:
        return Array.isArray(value) ? value : [value];
      default:
        return value;
    }
  };

  async function aggregateCollection(this: Resource<TSchema>, req: Request, res: Response) {
    const model = this.getModelFromRequest(req);
    // Prepare operation params
    const pipeline: AggregationPipeline = castQueryParam(
      parseQueryParam(req.query[pipelineParamName], pipelineParamName),
      pipelineParamName
    );
    const filter: FilterQuery<TSchema> = await this.buildRequestFilter(req);
    const options: CollectionAggregationOptions = {};
    const operation = createOperationMap<TSchema>({
      method: 'aggregateCollection',
      scope: 'collection',
      request: req,
      pipeline,
      filter
    });
    // Execute preHooks
    await this.hooks.execPre('filter', [operation, filter]);
    if (!isEmpty(operation.get('filter'))) {
      operation.set(
        'pipeline',
        ([{$match: operation.get('filter')}] as AggregationPipeline).concat(operation.get('pipeline'))
      );
    }
    await this.hooks.execPre('aggregateCollection', [operation, pipeline, options]);
    // Actual mongo call
    const result = await model.aggregate(operation.get('pipeline'), options);
    operation.set('result', result);
    // Execute postHooks
    await this.hooks.execPost('aggregateCollection', [operation, operation.get('pipeline'), options]);
    res.json(operation.get('result'));
  }

  resource.pre('buildPath', (router: Router, path: string) => {
    router.get(`${path}/${pathName}`, asyncHandler(aggregateCollection.bind(resource)));
  });
};
