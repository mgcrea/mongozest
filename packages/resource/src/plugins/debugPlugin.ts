import { DefaultSchema } from '@mongozest/core';
import { ResourceHookName } from '../typings';
import { OperationMap } from '../operation';
import { Resource } from '../resource';
import { inspect, log } from '../utils/logger';

export const debugPlugin = <TSchema extends DefaultSchema = DefaultSchema>(resource: Resource<TSchema>): void => {
  const logMethodOperation = (method: string, operation: OperationMap<TSchema>, ...args: any[]) => {
    if (operation.get('method') !== method) {
      return;
    }
    log(`resource.${method}(${args.map(inspect).join(', ')})`);
  };

  const loggedPreHooks: ResourceHookName[] = [
    'getCollection',
    'postCollection',
    'patchCollection',
    'deleteCollection',
    'getDocument',
    'patchDocument',
    'deleteDocument',
    'aggregateCollection',
  ];
  loggedPreHooks.forEach((name) => {
    resource.pre(name, (operation: OperationMap<TSchema>, ...args: unknown[]) => {
      logMethodOperation(name, operation, ...args);
    });
  });
};
