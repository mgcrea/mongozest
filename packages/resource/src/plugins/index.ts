import debugPlugin from './debugPlugin';
import createdByPlugin from './createdByPlugin';
import queryPlugin from './queryPlugin';
import populatePlugin from './populatePlugin';
import shortIdPlugin from './shortIdPlugin';

export {debugPlugin, createdByPlugin, queryPlugin, populatePlugin, shortIdPlugin};
