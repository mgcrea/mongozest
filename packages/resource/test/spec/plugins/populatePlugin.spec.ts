import {Model} from '@mongozest/core';
import {populatePlugin as modelPopulatePlugin} from '@mongozest/plugins';
import {omit} from 'lodash';
import {getDbName} from 'root/test/utils';
import createResource, {Resource} from 'src/index';
import populatePlugin from 'src/plugins/populatePlugin';
import {makeFetch} from 'supertest-fetch';
import {breakdownMiddleware, createTestApp} from 'test/utils/app';
import fixtures from 'test/utils/fixtures';

const DB_NAME = getDbName(__filename);

const app = createTestApp({routers: []});
const {mongo, redis, insertFixture} = app.locals;
app.locals.fixtures = fixtures;
const fetch = makeFetch(app);

class User extends Model {
  static schema = {
    firstName: {bsonType: 'string'},
    lastName: {bsonType: 'string'},
    email: {bsonType: 'string', required: true},
    nationality: {bsonType: 'string'},
    device: {bsonType: 'objectId'}
  };
}

class Comment extends Model {
  static schema = {
    text: {bsonType: 'string'},
    user: {bsonType: 'objectId', ref: 'User'}
  };
  static plugins = [modelPopulatePlugin];
}

beforeAll(async () => {
  const db = await mongo.connect(DB_NAME);
  await db.dropDatabase();
  await mongo.loadModels({User, Comment});
});

afterAll(async () => {
  await Promise.all([app.close()]);
});

describe('populatePlugin', () => {
  describe('resources', () => {
    describe('User', () => {
      let resource: Resource;
      it('should properly create resources', async () => {
        resource = createResource('User', {db: 'mongo'});
        expect(resource instanceof Resource).toBeTruthy();
      });
      it('should properly serve resource', async () => {
        const router = resource.build();
        app.use(router);
      });
    });
    describe('Comment', () => {
      let resource: Resource;
      it('should properly create resources', async () => {
        resource = createResource('Comment', {db: 'mongo', plugins: [populatePlugin]});
        expect(resource instanceof Resource).toBeTruthy();
      });
      it('should properly serve resource', async () => {
        const router = resource.build();
        app.use(router);
      });
    });
    afterAll(() => {
      app.use(breakdownMiddleware);
    });
  });
  describe('collection', () => {
    describe('GET /comments', () => {
      it('should return 200', async () => {
        const {_id: userId} = await insertFixture('user.mongozest');
        const {insertedId} = await mongo.model('Comment').insertOne({text: 'Hello World', user: userId});
        const query = `population=${JSON.stringify({user: 1})}`;
        const res = await fetch(`/comments?${query}`, {
          method: 'get',
          headers: {'Content-Type': 'application/json'}
        })
          .expect(200)
          .expect('content-type', /^application\/json/);
        const resBody = await res.json();
        expect(Array.isArray(resBody)).toBeTruthy();
        expect(resBody.length).toEqual(1);
        expect(omit(resBody[0], '_id', 'user')).toMatchSnapshot();
        expect(typeof resBody[0].user).toBe('object');
        expect(omit(resBody[0].user, '_id')).toMatchSnapshot();
      });
    });
  });
});
