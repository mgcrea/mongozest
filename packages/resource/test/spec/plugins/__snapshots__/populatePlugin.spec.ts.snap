// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`populatePlugin collection GET /comments should return 200 1`] = `
Object {
  "text": "Hello World",
}
`;

exports[`populatePlugin collection GET /comments should return 200 2`] = `
Object {
  "device": "1a1a836f1cf9087b520148de",
  "email": "john.doe@gmail.com",
  "firstName": "John",
  "lastName": "Doe",
}
`;
